### ISSUE TITLE
1.4.0: RWLockedGraph is not really thread-safe

### ISSUE THREAD

**Issue by sszuev** [avicomp/ont-api#65]
it is time to close this gap (before 1.4.0 release)

Unfortunately it is not straightforward: for a `ExtendedIterator` that is not properly closed and not exhausted I see only the one way: the use of in-memory cache. 
But for other cases - when iterator is closed or has reached its end - there is a quite simple solution: just control its end.  
So the final solution can be complex - include both (different) ways to ensure full thread-safety.

**Comment by sszuev** [comment]
The provided patch contains a mechanism that ensures thread-safety on iterators, but there is still no guarantee in model level, since, e.g., to retrieve all `OntClass`es it is need first get all root triples, and than construct  `OntClass` instances with found root triple, while this triple may disappear due to deleting by other thread.

So, in additional to `RWLockedGraph` need to introduce a new `RWLockedModel` (precisely speaking, it is need to move already existing anonymous model from the `ru.avicomp.ontapi.OntologyModelImpl.Concurrent` class up to jena package level). 
Or fix the existing `OntGraphModelImpl` to work with `RWLockedGraph`.

In other words, one more issue and corresponding commit are required. But, maybe, not in **1.4.0**

**Comment by sszuev** [comment]
done. close.
